### PR TITLE
Phase 2 + Zexe Re-implementation

### PR DESCRIPTION
**Pull request by gakonst**
As title.

- Tests for generating/contributing/verifying have been added

- A Groth16 E2E test is now included with a simple knowledge of square root circuit

- New module: `polynomial.rs`, contains all the necessary functions/tests for evaluating the polynomials.

- Removed all helper modules since we can use the shared test helpers for them.

- The ceremony has been re-implemented, IMO it's easier to read and understand now, as well as has a lot of extra tests. I recommend reading the new implementation with the [MMORPG paper](https://eprint.iacr.org/2017/1050) on the side when reviewing.

- Removed all the binaries. The consumer should import phase2 and use it with their own Circuit for their CLI. We _could_ add this back, if there was a good way to serialize/deserialize circuits.

Note: The `cs_hash` is _NOT_ correctly implemented right now, because `Parameters::write` is still unimplemented. Blocked on it until https://github.com/scipr-lab/zexe/pull/109 is merged.

### CODE REVIEW

File: phase2/src/parameters.rs [hunk 1]
```diff
@@ -658,188 +286,232 @@ impl MPCParameters {
...
-    }
+    Ok(result)
+}
 
+#[allow(unused)]
+fn hash_params<E: PairingEngine>(params: &Parameters<E>) -> Result<[u8; 64]> {
     let sink = io::sink();
     let mut sink = HashWriter::new(sink);
-    sink.write_all(&before.cs_hash[..]).unwrap();
+    // TODO: Re-enable this
```
Comment (gakonst) [inline phase2/src/parameters.rs:368]: This should be re-enabled once the Zexe's serialization supports writing the params.

File: phase2/src/parameters.rs [hunk 2]
```diff
@@ -385,239 +125,127 @@ impl MPCParameters {
...
+        // generate the transcript from the current contributions and the previous cs_hash
+        verify_transcript(before.cs_hash, &after.contributions)
     }
 
     /// Serialize these parameters. The serialized parameters
-    /// can be read by bellman as Groth16 `Parameters`.
-    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+    /// can be read by Zexe's Groth16 `Parameters`.
+    pub fn write<W: Write>(&self, mut writer: W) -> Result<()> {
+        // TODO: This is unimplemented
```
Comment (kobigurk) [inline phase2/src/parameters.rs:248]: Waiting for the serialization PR?

File: phase2/src/parameters.rs [hunk 3]
```diff
@@ -658,188 +286,232 @@ impl MPCParameters {
...
-        .unwrap();
-    sink.write_all(pubkey.s_delta.into_uncompressed().as_ref())
-        .unwrap();
+    // Allocate the "one" input variable
+    assembly.alloc_input(|| "", || Ok(E::Fr::one()))?;
+    // Synthesize the circuit.
+    circuit.generate_constraints(&mut assembly)?;
+    // Reject circuits larger than 2^28
+    let exp = log_2(assembly.num_constraints);
+    if exp > 28 {
```
Comment (kobigurk) [inline phase2/src/parameters.rs]: No need to reject this anymore, especially untrue for Zexe curves

**Review by kobigurk — APPROVED** [review]

File: phase2/src/parameters.rs (hunk 2 above)
Comment (gakonst) [reply phase2/src/parameters.rs:248]: Yep.

**Comment by Pratyush** [comment]
https://github.com/scipr-lab/zexe/pull/109 is now merged =)

**Pull request merged**